kiwmi/desktop/output.c

```diff
@@ -376,8 +386,129 @@ new_output_notify(struct wl_listener *listener, void *data)
+        head->state.mode    = output->wlr_output->current_mode;
+        if (NULL != area) {
+            head->state.x = area->x;
+            head->state.y = area->y;
+        }
+    }
+
+    wlr_output_manager_v1_set_configuration(
+        desktop->output_manager, configuration);
+}
+
+static void
+output_manager_configure(
+    struct kiwmi_server *server,
+    struct wlr_output_configuration_v1 *configuration,
+    bool test)
+{
+    bool okay = true;
+    struct wlr_output_configuration_head_v1 *head;
+    wl_list_for_each (head, &configuration->heads, link) {
+        struct wlr_output *wlr_output = head->state.output;
+        struct kiwmi_output *output   = wlr_output->data;
+
+        output->enable_changed = head->state.enabled ^ wlr_output->enabled;
+
+        wlr_output_enable(wlr_output, head->state.enabled);
+        if (head->state.enabled) {
+            if (head->state.mode) {
+                wlr_output_set_mode(wlr_output, head->state.mode);
+            } else {
+                wlr_output_set_custom_mode(
+                    wlr_output,
+                    head->state.custom_mode.width,
+                    head->state.custom_mode.height,
+                    head->state.custom_mode.refresh);
+            }
+
+            wlr_output_layout_move(
+                server->desktop.output_layout,
+                wlr_output,
+                head->state.x,
+                head->state.y);
+            wlr_output_set_transform(wlr_output, head->state.transform);
+            wlr_output_set_scale(wlr_output, head->state.scale);
+        }
+
+        if (!wlr_output_test(wlr_output)) {
+            okay = false;
+            break;
+        }
+    }
+
+    wl_list_for_each (head, &configuration->heads, link) {
+        if (okay && !test) {
+            struct kiwmi_output *output = head->state.output->data;
+            if (head->state.enabled) {
+                if (output->enable_changed) {
+                    wl_signal_emit(&output->desktop->events.new_output, output);
+                } else {
+                    wl_signal_emit(&output->events.resize, output);
+                }
+            } else if (output->enable_changed) {
+                wl_signal_emit(&output->events.destroy, output);
+            }
+            wlr_output_commit(head->state.output);
+        } else {
+            wlr_output_rollback(head->state.output);
+        }
+    }
+
+    if (okay) {
+        wlr_output_configuration_v1_send_succeeded(configuration);
+    } else {
+        wlr_output_configuration_v1_send_failed(configuration);
+    }
+
+    wlr_output_configuration_v1_destroy(configuration);
+}
+
+void
+output_manager_apply_notify(struct wl_listener *listener, void *data)
+{
+    struct kiwmi_desktop *desktop =
+        wl_container_of(listener, desktop, output_manager_apply);
+    struct kiwmi_server *server = wl_container_of(desktop, server, desktop);
+    struct wlr_output_configuration_v1 *configuration = data;
```

Comment:
Again i think i’d change the order (also below), but this is quite a weird situation in which i’m unsure.

Reply:
This one, I've left because they're used in the order sever then configuration, and you need to get desktop to get server.

But if anyone has a particularly strong opinion, I'll happily change it.